File: carddemo-java/src/main/java/com/cardemo/entity/Account.java
package com.cardemo.entity;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;

/**
 * JPA entity mapped from COBOL copybook CVACT01Y.cpy (ACCOUNT-RECORD).
 * Total COBOL record length: 300 bytes.
 * Seed data file: acctdata.txt (50 records).
 */
@Entity
@Table(name = "accounts")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Account {

    /** ACCT-ID — PIC 9(11), bytes [0:11]. Primary key. */
    @Id
    @Column(name = "acct_id")
    private Long acctId;

    /** ACCT-ACTIVE-STATUS — PIC X(01), bytes [11:12]. Active flag (Y/N). */
    @Column(name = "acct_active_status", length = 1)
    private String acctActiveStatus;

    /** ACCT-CURR-BAL — PIC S9(10)V99, bytes [12:24]. Current balance (zoned decimal). */
    @Column(name = "acct_curr_bal", precision = 12, scale = 2)
    private BigDecimal acctCurrBal;

    /** ACCT-CREDIT-LIMIT — PIC S9(10)V99, bytes [24:36]. Credit limit (zoned decimal). */
    @Column(name = "acct_credit_limit", precision = 12, scale = 2)
    private BigDecimal acctCreditLimit;

    /** ACCT-CASH-CREDIT-LIMIT — PIC S9(10)V99, bytes [36:48]. Cash credit limit (zoned decimal). */
    @Column(name = "acct_cash_credit_limit", precision = 12, scale = 2)
    private BigDecimal acctCashCreditLimit;

    /** ACCT-OPEN-DATE — PIC X(10), bytes [48:58]. Account open date (YYYY-MM-DD). */
    @Column(name = "acct_open_date")
    private LocalDate acctOpenDate;

    /** ACCT-EXPIRAION-DATE — PIC X(10), bytes [58:68]. Expiration date (YYYY-MM-DD). */
    @Column(name = "acct_expiration_date")
    private LocalDate acctExpirationDate;

    /** ACCT-REISSUE-DATE — PIC X(10), bytes [68:78]. Reissue date (YYYY-MM-DD). */
    @Column(name = "acct_reissue_date")
    private LocalDate acctReissueDate;

    /** ACCT-CURR-CYC-CREDIT — PIC S9(10)V99, bytes [78:90]. Current cycle credit (zoned decimal). */
    @Column(name = "acct_curr_cyc_credit", precision = 12, scale = 2)
    private BigDecimal acctCurrCycCredit;

    /** ACCT-CURR-CYC-DEBIT — PIC S9(10)V99, bytes [90:102]. Current cycle debit (zoned decimal). */
    @Column(name = "acct_curr_cyc_debit", precision = 12, scale = 2)
    private BigDecimal acctCurrCycDebit;

    /** ACCT-ADDR-ZIP — PIC X(10), bytes [102:112]. Address ZIP code. */
    @Column(name = "acct_addr_zip", length = 10)
    private String acctAddrZip;

    /** ACCT-GROUP-ID — PIC X(10), bytes [112:122]. Account group identifier. */
    @Column(name = "acct_group_id", length = 10)
    private String acctGroupId;

    // FILLER — PIC X(178), bytes [122:300]. Padding — not mapped.
}

File: app/python-api/app/models/account.py
"""Pydantic model for ACCT-RECORD (CVACT01Y.cpy — 300 bytes).

COBOL layout:
  01 ACCT-RECORD.
    05 ACCT-ID                PIC 9(11).        -- bytes 1-11
    05 ACCT-ACTIVE-STATUS     PIC X(01).        -- byte 12 (Y/N)
    05 ACCT-CURR-BAL          PIC S9(10)V99.    -- BigDecimal scale 2
    05 ACCT-CREDIT-LIMIT      PIC S9(10)V99.
    05 ACCT-CASH-CREDIT-LIMIT PIC S9(10)V99.
    05 ACCT-OPEN-DATE         PIC X(10).        -- YYYY-MM-DD
    05 ACCT-EXPIRAION-DATE    PIC X(10).
    05 ACCT-REISSUE-DATE      PIC X(10).
    05 ACCT-CURR-CYC-CREDIT   PIC S9(10)V99.
    05 ACCT-CURR-CYC-DEBIT    PIC S9(10)V99.
    05 ACCT-ADDR-ZIP          PIC X(10).
    05 ACCT-GROUP-ID          PIC X(10).
    05 FILLER                 PIC X(178).
"""

from decimal import Decimal

from pydantic import BaseModel, Field


class AccountRecord(BaseModel):
    """Account master record — maps to CVACT01Y.cpy (300-byte ACCT-RECORD)."""

    acct_id: str = Field(
        ...,
        min_length=11,
        max_length=11,
        pattern=r"^\d{11}$",
        description="Account ID — 11-digit numeric string",
    )
    acct_active_status: str = Field(
        ...,
        min_length=1,
        max_length=1,
        description="Active status flag (Y/N)",
    )
    acct_curr_bal: Decimal = Field(
        ...,
        max_digits=12,
        decimal_places=2,
        description="Current balance — PIC S9(10)V99",
    )
    acct_credit_limit: Decimal = Field(
        ...,
        max_digits=12,
        decimal_places=2,
        description="Credit limit — PIC S9(10)V99",
    )
    acct_cash_credit_limit: Decimal = Field(
        ...,
        max_digits=12,
        decimal_places=2,
        description="Cash credit limit — PIC S9(10)V99",
    )
    acct_open_date: str = Field(
        ...,
        max_length=10,
        description="Account open date — YYYY-MM-DD",
    )
    acct_expiration_date: str = Field(
        ...,
        max_length=10,
        description="Account expiration date — YYYY-MM-DD",
    )
    acct_reissue_date: str = Field(
        ...,
        max_length=10,
        description="Reissue date — YYYY-MM-DD",
    )
    acct_curr_cyc_credit: Decimal = Field(
        ...,
        max_digits=12,
        decimal_places=2,
        description="Current cycle credit — PIC S9(10)V99",
    )
    acct_curr_cyc_debit: Decimal = Field(
        ...,
        max_digits=12,
        decimal_places=2,
        description="Current cycle debit — PIC S9(10)V99",
    )
    acct_addr_zip: str = Field(
        ...,
        max_length=10,
        description="Address ZIP code",
    )
    acct_group_id: str = Field(
        ...,
        max_length=10,
        description="Group ID",
    )
